mtwitter.js

```diff
@@ -0,0 +1,254 @@
+                  cb(null);
+              }
+          } catch (e) {
+              console.error("Error getting bearer token:", e);
+              cb(null);
+          }
+
+      }
+  });
+}
+
+mtwitter.prototype.get_user_by_id = function(id) {
+    var ts = new Date().getTime();
+    return new Promise((resolve, reject) => {
+        for (var name in this.tuser_cache) {
+            var cached = this.tuser_cache[name];
+            if (ts - cached.cache_time > TWITTER_PROFILE_INTERVAL) {
+                continue;
+            }
+            if (cached.user != null && cached.user.id_str == id) {
+                resolve(cached.user);
+            }
+        }
+        this.app_twitter.get('users/show', {
+            user_id: id
+        }, (error, user, response) => {
+            if (error) {
+                console.error(error);
+                reject(error);
+            } else {
+                this.tuser_cache[user.screen_name] = {
+                    cache_time: ts,
+                    user: user
+                };
+                resolve(user);
+            }
+        });
+    });
+}
+
+mtwitter.prototype.get_user = function(name) {
+    console.log("Looking up @" + name);
+    return new Promise((resolve, reject) => {
+        var ts = new Date().getTime();
+        if (this.tuser_cache[name] != undefined) {
+            console.log("Checking cache for @" + name);
+            var cached = this.tuser_cache[name];
+            if (ts - cached.cache_time < TWITTER_PROFILE_INTERVAL) {
+                resolve(cached.user);
+            }
+        }
+
+        console.log("Checking twitter for @" + name);
+        this.app_twitter.get('users/show', {
+            screen_name: name
+        }, (error, user, response) => {
+            if (error) {
+                this.tuser_cache[name] = {
+                    cache_time: ts,
+                    user: null
+                };
+                reject(error);
+            }
+            this.tuser_cache[name] = {
+                cache_time: ts,
+                user: user
+            };
+            resolve(user);
+        });
+    });
+
+}
+
+/*
+  Add a user's timeline to the iterator. Tweets will be automatically send to
+  the given room.
+*/
+mtwitter.prototype.enqueue_timeline = function(userid, localroom, remoteroom) {
+    var obj = {
+        "user_id": userid,
+        "local": localroom,
+        "remote": remoteroom
+    };
+    this.timeline_list.push(obj);
+    this.timeline_queue.push(obj);
+}
+
+/*
+  This function will fill the content structure for a new matrix message
+  for a given tweet.
+*/
+mtwitter.prototype.construct_message = function(tweet, type) {
+    return {
+        "body": tweet.text,
+        "created_at": tweet.created_at,
+        "likes": tweet.favorite_count,
+        "reblogs": tweet.retweet_count,
+        "tweet_id": tweet.id_str,
+        "tags": tweet.entities.hashtags,
```

Comment:
> start validating all the structures I receive, which feels a bit overkill.

It's never overkill to sanity check what you're getting from the wire. It's more lines of code, it's more faff to do, it takes fractionally longer to process messages, but you'll then have some guarantees about the data you're manipulating. I appreciate it *is* annoying to do though, so I don't mind whether you do or do not. However, you should be under no illusion that doing it is in some way pointless/not worth the effort.

What's more, it becomes clearer to outsiders visiting your code as they can see "oh this takes a tweet, which I can see in file X has these properties", which saves them the effort of googling the Twitter API to find the JSON response format to the call you're making. You do run the risk of redocumenting the Twitter API in your code if you take this to the extreme though, so a balance should be struck. In other languages (e.g. Go), you basically decode the JSON into an instance of a class (aka `struct`) which defines exactly what you **think** you should be getting from the wire. Checking those assumptions is never overkill.

If you *do* decide to go down this route, a fairly noddy `Tweet` class would serve to at least assert the object form:

```javascript
class Tweet {
  constructor(tweetJson) {
    tweetJson.entities = tweetJson.entities || []; // default `entities` to an empty array if the property is missing
    tweetJson.favorite_count = tweetJson.favorite_count || 0; // default to 0
    // etc..
    this.data = tweetJson;
  }
}

t = new Tweet(someJsonObj);
// e.g. access via t.data.entities
```